app/use_cases/create_or_update_protocol.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+class CreateOrUpdateProtocol < ActiveInteraction::Base
+  string :name
+  integer :duration
+  string :invitation_text
+  string :informed_consent_questionnaire_key, default: nil
+
+  attr_reader :protocol
+
+  array :questionnaires do
+    hash do
+      string :key
+      hash :measurement do
+        integer :open_from_offset
+        string :open_from_day, default: nil
+        integer :period, default: nil
+        integer :open_duration, default: nil
+        integer :reminder_delay, default: nil
+        integer :priority
+        boolean :stop_measurement
+        boolean :should_invite
+        boolean :only_redirect_if_nothing_else_ready
+      end
+    end
+  end
+
+  validates :name, presence: true
+  validates :duration, presence: true
+  validates :invitation_text, presence: true
```

Comment:
Why are all these validations needed? Doesn't just having a property put that validation in there and also the error messages?

For example, if I look at the `CalculateScores` use case, which requires two arguments, a `questionnaire` hash an a `content` hash, then if I leave out the `questionnaire` hash, I get the following error when calling `.run!`:
![ConEmu64_2022-04-06_17-45-27](https://user-images.githubusercontent.com/607965/162014991-84b89da5-e304-44a3-825e-ff0495a6247c.png)

and if I leave out exclamation point and just collect the .errors array, then I just get the error message for the attribute that I missed.

So why the `validates` statements?

Reply:
you are right, the mere fact of declaring the properties indicates their presence is needed. TIL.